Service worker: Tests for updateViaCache (previously useCache)

jakearchibald opened the pull request:
<!-- Reviewable:start -->

<!-- Reviewable:end -->

jakearchibald committed:
Tests for updateViaCache (previously useCache)

jakearchibald commented:
+@wanderview

jakearchibald commented:
Generating the tests in a loop seemed like the simplest way forward here. Meant I could write the tests fairly linearly without abstracting a lot. But I can look at changing it if it's too confusing.

Also, yay async functions!

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/219977292)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 5f28375c18930e871e863b4fa64c241420f8ed7f
Using browser at version BuildID 20170407100252; SourceStamp 10ea10d9993c9701e5525928257a589dea2c05d8
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/5515/service-workers/service-worker/registration-updateviacache.https.html">/service-workers/service-worker/registration-updateviacache.https.html</a></summary>


|                         Subtest                         | Results |                           Messages                           |
|---------------------------------------------------------|---------|--------------------------------------------------------------|
|                                                         | OK      |                                                              |
| `register-with-updateViaCache-undefined`                | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports`                  | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all`                      | PASS    |                                                              |
| `register-with-updateViaCache-never`                    | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-undefined` | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-imports`   | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-all`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-never`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-undefined`   | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-imports`     | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-all`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-never`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-undefined`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-imports`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-all`             | PASS    |                                                              |
| `register-with-updateViaCache-all-then-never`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-undefined`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-imports`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-all`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-never`         | PASS    |                                                              |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/219977293)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 5f28375c18930e871e863b4fa64c241420f8ed7f
Using browser at version 59.0.3063.4 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/5515/service-workers/service-worker/registration-updateviacache.https.html">/service-workers/service-worker/registration-updateviacache.https.html</a></summary>


|                         Subtest                         | Results |                           Messages                           |
|---------------------------------------------------------|---------|--------------------------------------------------------------|
|                                                         | OK      |                                                              |
| `register-with-updateViaCache-undefined`                | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports`                  | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all`                      | PASS    |                                                              |
| `register-with-updateViaCache-never`                    | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-undefined` | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-imports`   | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-all`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-never`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-undefined`   | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-imports`     | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-all`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-never`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-undefined`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-imports`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-all`             | PASS    |                                                              |
| `register-with-updateViaCache-all-then-never`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-undefined`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-imports`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-all`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-never`         | PASS    |                                                              |

</details>
</details>

jakearchibald committed:
Fixing trailing whitespace

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5515/)

jakearchibald commented:
The failing tests (in Chrome & Firefox) are as-expected.

wanderview commented:
We actually have a wpt for the cache changes written, but not uplifted yet.  Sorry if there was duplicate work making this test.  We can either keep both or just this one.  I'll talk to Ho-Pang about it Monday.

jakearchibald commented:
Ah, sorry! Anyway, I'll let you pick whichever & I promise not to be offended. 😆

wanderview commented:
No problem!  The test here has the benefit of being run against chrome to check for flakiness.  Our wpt uplift process is lacking that step.

hopanghsu commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,150 @@
+      await reg.update();
+
+      if (updateViaCache == 'all') {
+        assert_equals(sw.installing, undefined, "No new service worker");
```
By `sw.installing`, are we referring to `reg.installing`?

jakearchibald replied:
Whoops! Good catch.

hopanghsu reviewed (commented):
Thanks Jake,

The code is arranged elegantly! I left some questions in the comments. In addition to those, I think it might be a good idea to check whether `updatefound` is dispatched when there should be a new service worker, but I have no strong preference on this.

jakearchibald commented:
@hopanghsu what's the benefit of checking `updatefound` here? The test already checks if there's been an update.

jakearchibald committed:
Was checking .installing on wrong object.

wanderview commented:
> what's the benefit of checking updatefound here?

Just because the spec requires firing one?  Maybe that's covered by a different test.

jakearchibald commented:
@wanderview yeah, my feeling is it's unrelated to this test.

jakearchibald commented:
I've resolved the merge conflicts. Are we good to merge this?

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231757879)

# Firefox (nightly) #
Testing web-platform-tests at revision bc4f24fba23f0dd0f1c19a23a77affa1ccfd9a82
Using browser at version BuildID 20170512100218; SourceStamp 8a7d0b15595f9916123848ca906f29c62d4914c9
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/5515/service-workers/service-worker/registration-updateviacache.https.html">/service-workers/service-worker/registration-updateviacache.https.html</a></summary>


|                         Subtest                         | Results |                           Messages                           |
|---------------------------------------------------------|---------|--------------------------------------------------------------|
|                                                         | OK      |                                                              |
| `register-with-updateViaCache-undefined`                | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports`                  | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all`                      | PASS    |                                                              |
| `register-with-updateViaCache-never`                    | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-undefined` | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-imports`   | PASS    |                                                              |
| `register-with-updateViaCache-undefined-then-all`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-undefined-then-never`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-undefined`   | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-imports`     | PASS    |                                                              |
| `register-with-updateViaCache-imports-then-all`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-imports-then-never`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-undefined`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-imports`         | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-all-then-all`             | PASS    |                                                              |
| `register-with-updateViaCache-all-then-never`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-undefined`     | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-imports`       | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-all`           | FAIL    | `assert_true: New worker installing expected true got false` |
| `register-with-updateViaCache-never-then-never`         | PASS    |                                                              |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231757885)

# Chrome (unstable) #
Testing web-platform-tests at revision bc4f24fba23f0dd0f1c19a23a77affa1ccfd9a82

jakearchibald committed:
“never” to “none”

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239594339)

# Sauce (safari) #
Testing web-platform-tests at revision 1164ee652bfd40ce0e4f4348716be80a2799c5ea
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/5515/service-workers/service-worker/registration-updateviacache.https.html">/service-workers/service-worker/registration-updateviacache.https.html</a></summary>


|                    Subtest                    | Results |                   Messages                   |
|-----------------------------------------------|---------|----------------------------------------------|
|                                               | OK      |                                              |
| `Service Worker: Registration-updateViaCache` | FAIL    | `SyntaxError: Unexpected keyword 'function'` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239594348)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 1164ee652bfd40ce0e4f4348716be80a2799c5ea
Using browser at version 14.14393

jakearchibald committed:
Testing setting via link element & link header

domenic commented on html/dom/elements-metadata.js:
```diff
@@ -18,6 +18,11 @@ var metadataElements = {
+    scope: "string",
+    updateViaCache: {
+      type: "enum",
+      keywords: ["imports", "all", "none"]
```
You'll want to add `defaultVal` here too.

This is used by http://w3c-test.org/html/dom/reflection-metadata.html

domenic reviewed (commented):
I'd rather an implementer (like @mkruisselbrink?) review the actual processing model parts of the tests, but only one minor comment on the reflection part.

jakearchibald committed:
Adding default & invalid val

domenic commented:
Reflection tests LGTM but would like LGTM from someone else on the actual semantics tests.

mattto commented:
Thanks for the tests! The discussion here seems relevant: https://groups.google.com/a/chromium.org/d/msg/net-dev/wbhbSn9ICfc/XDmCkpboAAAJ

The conclusion there was that browser caching behaviors cannot be tested in WPTs, because caching is a SHOULD.

wanderview commented:
> The conclusion there was that browser caching behaviors cannot be tested in WPTs, because caching is a SHOULD.

Is there an actual compat issue that we expect to prevent these tests passing in currently implementing browsers?

I would prefer to keep the tests and if a browser wants to deviate from the SHOULD language they can internally mark the test as disabled/expected-fail/etc.

mattto commented:
OK. It makes sense to me that WPT tests can test SHOULDs. Sounds good.

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+    return new Promise(r => setTimeout(r, ms));
+  }
+
+  function waitForActivated(sw) {
```
This could use wait_for_state(t, sw, 'activated')

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+<script>
+  const updateViaCacheValues = [undefined, 'imports', 'all', 'none'];
+  const scriptUrl = 'resources/update-max-aged-worker.py';
+  const scope = 'resources/blank.html';
```
To make clear they are global constants, could we name these globals like kUpdateViaCacheValues, kScope or else UPDATE_VIA_CACHE_VALUES, SCOPE, etc?

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+    });
+  }
+
+  function registerViaApi(scriptUrl, opts) {
```
A little confusing that this param is the same name as the global |scriptUrl|.

jakearchibald replied:
This was fixed by renaming the global

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+        );
+
+        const sw = reg.installing || reg.waiting || reg.active;
+        if (!sw) console.log(reg);
```
Do you need this console.log()? It tends to pollute test results. Should it just be an assert() instead?

jakearchibald replied:
Whoops!

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+            assert_not_equals(values.mainTime, newValues.mainTime, "Main script should have updated");
+            assert_equals(values.importTime, newValues.importTime, "Imported script should be the same");
+          }
+          else { // 'none'
```
To be more clear: // updateViaCache == 'none'

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+            assert_not_equals(values.mainTime, newValues.mainTime, "Main script should have updated");
+            assert_equals(values.importTime, newValues.importTime, "Imported script should be the same");
+          }
+          else { // 'none'
```
same as line 167

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+    const reg = await navigator.serviceWorker.getRegistration(scope);
+    if (!reg) return;
+    if (reg.scope == new URL(scope, location).href) {
+      return reg.unregister()
```
nit: add a semi-colon

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+        return reg;
+      }
+
+      await wait(100);
```
The polling with timeout is a bit unfortunate, it puts a bound on how fast the test can run (100ms per test adds up for hundreds of tests). A similar test, register-link-element.https.html, just waits for link.onload. Could we just do that here?

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+        return reg;
+      }
+
+      await wait(100);
```
register-link-header.https.html uses an iframe and navigator.serviceWorker.ready. Could we do that here instead?

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+
+        await navigator.serviceWorker.register(fullScriptUrl, opts);
+
+        // If there's no change, register should be a no-op.
```
This was a bit subtle for me. Recommend saying "If there's no change to the updateViaCache value"

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+    }
+  }
+
+  // Testing changing registration
```
// Test changing the updateViaCache value of an existing registration.

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,230 @@
+    [registerViaLinkHeader, 'via-link-header']
+  ];
+
+  // Testing creating registrations
```
service worker WPT tests tend to follow Chromium Style Guide for comments, so this would take the form:
// Test creating registrations.
(imperative form + full stop)

jakearchibald replied:
TIL, thanks!

mattto commented:
Jake do you want to continue with this PR? Is it ok if we take what's here now and work on it in a Chromium codereview as we're trying to implement?

jakearchibald commented:
@mattto I'll address your comments now. Apologies for the (usual) slowness, was out conferencing last week.

jakearchibald committed:
Responding to feedback

jakearchibald commented:
@mattto Done!

mattto commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,210 @@
+    const fullScope = new URL(opts.scope, window.location).href;
+    scriptUrl = new URL(scriptUrl, location).href;
+
+    // Assuming that if this isn't supported, the header isn't
```
nit: Sort of looks like the sentence got cut-off mid-sentence. "Assume that if the link element doesn't support serviceworker, the header doesn't either." ?

jakearchibald replied:
Oh wow, now you point it out, that comment was really badly written. Fixed with your suggestion.

mfalken reviewed (approved):
Nice test! lgtm, thanks!

jakearchibald committed:
Improving comment wording

mattto commented:
Thanks! Can we get this reviewed and committed so our implementation can use it (@mkruisselbrink)?

There is a linter error above:
service-workers/service-worker/registration-updateviacache.https.html:21: setTimeout used; step_timeout should typically be used instead (SET TIMEOUT)

hopanghsu reviewed (commented):
Hi all, and sorry for jumping in here. I am currently working on the related implementation, and I found something we might want to change.

hopanghsu commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,210 @@
+
+        // If there's no change to the updateViaCache value, register should be a no-op.
+        // The default value should behave as 'imports'.
+        if ((updateViaCache1 || 'imports') == (updateViaCache2 || 'imports')) {
```
Maybe we want to do this when (updateViaCache2 == 'all'), since it cannot pass the byte-check and thus there won't be a new service worker in that case.

jakearchibald replied:
Wow, good catch. Because we ignore the byte-check when the script URL changes, I thought we'd do the same here. But it doesn't make sense to, and we don't 😄

hopanghsu commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,210 @@
+            assert_equals(values.mainTime, newValues.mainTime, "Main script should be the same");
+            assert_equals(values.importTime, newValues.importTime, "Imported script should be the same");
+          }
+          else if (updateViaCache2 == 'imports') {
```
I think we should explicitly handle the default case, `updateViaCache2 == undefined`. Otherwise, it goes to the `else` branch.

hopanghsu commented on service-workers/service-worker/registration-updateviacache.https.html:
```diff
@@ -0,0 +1,210 @@
+          assert_true(!!newWorker, "New worker installing");
+          const newValues = await getScriptTimes(newWorker, testName);
+
+          if (updateViaCache2 == 'all') {
```
Since there shouldn't be a newly created service worker, we don't have to check anything here.

jakearchibald committed:
Fixing 'all' case. Also testing reg.updateViaCache.

jakearchibald commented:
@hopanghsu Don't apologise, thanks for the great feedback! Should be addressed now.

@mattto I've addressed that linting issue (the `setTimeout` was in a no-longer-used function anyway)

I've also added tests for `reg.updateViaCache`.

wanderview reviewed (approved):
rs+ based on previous review comments being addressed.
